snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,315 @@
+
+	ratio := float64(processingWeight) / float64(totalValidatorWeight)
+
+	if ratio > stakeThresholdForStragglerSuspicion {
+		sa.log.Trace("We are straggling behind", zap.Float64("ratio", ratio))
+		return true
+	}
+
+	sa.log.Trace("Nodes ahead of us:", zap.Float64("ratio", ratio))
+
+	return false
+}
+
+type snapshotter struct {
+	// log logs events
+	log logging.Logger
+
+	// minConfirmationThreshold is the minimum stake percentage that below it, we do not check if we are stragglers.
+	minConfirmationThreshold float64
+
+	// lastAccepted returns the last accepted block of this node.
+	lastAccepted func() ids.ID
+
+	// connectedValidators returns a set of tuples of NodeID and corresponding weight.
+	connectedValidators func() set.Set[ids.NodeWeight]
+
+	// lastAcceptedByNodeID returns the last accepted height a node has reported, or false if it is unknown.
+	lastAcceptedByNodeID func(id ids.NodeID) (ids.ID, bool)
+}
+
+func (s *snapshotter) validateNetInfo(netInfo netInfo) bool {
+	if netInfo.totalValidatorWeight == 0 {
+		s.log.Trace("Connected to zero weight")
+		return false
+	}
+
+	totalKnownLastBlockStakePercent := float64(netInfo.totalWeightWeKnowItsLastAcceptedBlock) / float64(netInfo.totalValidatorWeight)
+	stakeAheadOfUs := float64(netInfo.totalPendingStake) / float64(netInfo.totalValidatorWeight)
+
+	// Ensure we have collected last accepted blocks for at least 80% (or so) stake of the total weight we are connected to.
+	if totalKnownLastBlockStakePercent < minimumStakeThresholdRequiredForNetworkInfo {
+		s.log.Trace("Not collected enough information about last accepted blocks for the validators we are connected to",
+			zap.Float64("ratio", totalKnownLastBlockStakePercent))
+		return false
+	}
+
+	if stakeAheadOfUs < knownStakeThresholdRequiredForAnalysis {
+		s.log.Trace("Most stake we're connected to has the same height as we do",
+			zap.Float64("ratio", stakeAheadOfUs))
+		return false
+	}
+
+	return true
+}
+
+func (s *snapshotter) getNetworkSnapshot() (snapshot, bool) {
+	ourLastAcceptedBlock := s.lastAccepted()
+
+	netInfo, err := s.getNetworkInfo(ourLastAcceptedBlock)
+	if err != nil {
+		return snapshot{}, false
+	}
+
+	if !s.validateNetInfo(netInfo) {
+		return snapshot{}, false
+	}
+
+	snap := snapshot{
+		lastAcceptedBlockID:  netInfo.nodeWeightToLastAccepted,
+		totalValidatorWeight: netInfo.totalValidatorWeight,
+	}
+
+	return snap, true
+}
+
+func (s *snapshotter) getNetworkInfo(ourLastAcceptedBlock ids.ID) (netInfo, error) {
+	var res netInfo
+
+	validators := nodeWeights(s.connectedValidators().List())
+	nodeWeightToLastAccepted := make(nodeWeightsToBlocks, len(validators))
+
+	for _, vdr := range validators {
+		lastAccepted, ok := s.lastAcceptedByNodeID(vdr.ID)
+		if !ok {
+			continue
+		}
+		nodeWeightToLastAccepted[vdr] = lastAccepted
+	}
+
+	totalValidatorWeight, err := validators.totalWeight()
+	if err != nil {
+		s.log.Error("Failed computing total weight", zap.Error(err))
```

Comment:
Rather than emitting error logs in each place where we return an error, would it make sense to leave logging the error to the caller? Seems this is only called from one place.